File: pipecat_flutter_ios/ios/pipecat_flutter_ios/Sources/pipecat_flutter_ios/LocalMicStateController.swift
import Foundation

struct LocalMicSnapshot: Equatable {
  let inputEnabled: Bool
  let publishingEnabled: Bool

  var isSending: Bool {
    inputEnabled && publishingEnabled
  }

  func matchesDesired(_ enabled: Bool) -> Bool {
    inputEnabled == enabled && publishingEnabled == enabled
  }
}

struct LocalMicStateControllerConfig {
  let pollInterval: TimeInterval
  let maxPollAttempts: Int

  init(pollInterval: TimeInterval = 0.1, maxPollAttempts: Int = 10) {
    self.pollInterval = pollInterval
    self.maxPollAttempts = maxPollAttempts
  }
}

struct LocalMicStateControllerError: Error {
  let code: String
  let message: String
}

protocol LocalMicScheduledTask: AnyObject {
  func cancel()
}

@MainActor
protocol LocalMicStateControllerDelegate: AnyObject {
  func issueSetInputEnabled(
    _ enabled: Bool,
    completion: @escaping (Result<Void, Swift.Error>) -> Void
  )
  func issueSetPublishingEnabled(
    _ enabled: Bool,
    completion: @escaping (Result<Void, Swift.Error>) -> Void
  )
  func readSnapshot() -> LocalMicSnapshot?
  func schedule(delay: TimeInterval, action: @escaping () -> Void) -> LocalMicScheduledTask
}

@MainActor
final class LocalMicStateController {
  enum Phase {
    case idle
    case waitingForInputEnable
    case waitingForPublishEnable
    case waitingForPublishDisable
    case waitingForInputDisable
  }

  private struct Transition {
    let generation: Int64
    let phase: Phase
    let completion: ((Result<Void, Swift.Error>) -> Void)?
    let pollAttempts: Int
    let didReapply: Bool
  }

  private weak var delegate: LocalMicStateControllerDelegate?
  private let config: LocalMicStateControllerConfig

  private var generationCounter: Int64 = 0
  private var transition: Transition?
  private var pendingPoll: LocalMicScheduledTask?

  private(set) var desiredEnabled = true
  private(set) var observedSnapshot: LocalMicSnapshot?

  init(
    delegate: LocalMicStateControllerDelegate,
    config: LocalMicStateControllerConfig = LocalMicStateControllerConfig()
  ) {
    self.delegate = delegate
    self.config = config
  }

  var phase: Phase {
    transition?.phase ?? .idle
  }

  var isTransitioning: Bool {
    transition != nil
  }

  var isSendingAudio: Bool {
    observedSnapshot?.isSending == true
  }

  func reset(desiredEnabled: Bool = true) {
    self.desiredEnabled = desiredEnabled
    observedSnapshot = nil
    generationCounter += 1
    transition = nil
    cancelPendingPoll()
  }

  func updateDesiredEnabled(
    _ enabled: Bool,
    completion: ((Result<Void, Swift.Error>) -> Void)? = nil
  ) {
    desiredEnabled = enabled
    let snapshot = refreshSnapshot()
    if snapshot?.matchesDesired(enabled) == true {
      completeActiveTransition(with: .success(()))
      completion?(.success(()))
      return
    }

    cancelActiveTransition(with: .success(()))
    startTransition(snapshot: snapshot, completion: completion)
  }

  func observeSnapshot(_ snapshot: LocalMicSnapshot) {
    observedSnapshot = snapshot

    guard let activeTransition = transition else {
      if !snapshot.matchesDesired(desiredEnabled) {
        startTransition(snapshot: snapshot, completion: nil)
      }
      return
    }

    if snapshot.matchesDesired(desiredEnabled) {
      completeActiveTransition(with: .success(()))
      return
    }

    if !phaseSatisfied(activeTransition.phase, snapshot: snapshot) {
      return
    }

    let nextPhase = nextPhase(snapshot: snapshot, desiredEnabled: desiredEnabled)
    guard let nextPhase else {
      completeActiveTransition(with: .success(()))
      return
    }

    if nextPhase == activeTransition.phase {
      return
    }

    cancelPendingPoll()
    transition = Transition(
      generation: activeTransition.generation,
      phase: nextPhase,
      completion: activeTransition.completion,
      pollAttempts: 0,
      didReapply: false
    )
    executeCurrentPhase()
  }

  func currentMicEnabled(fallback: Bool) -> Bool {
    observedSnapshot?.isSending ?? fallback
  }

  private func startTransition(
    snapshot: LocalMicSnapshot?,
    completion: ((Result<Void, Swift.Error>) -> Void)?
  ) {
    guard let initialPhase = nextPhase(snapshot: snapshot, desiredEnabled: desiredEnabled) else {
      completion?(.success(()))
      return
    }

    transition = Transition(
      generation: nextGeneration(),
      phase: initialPhase,
      completion: completion,
      pollAttempts: 0,
      didReapply: false
    )
    executeCurrentPhase()
  }

  private func executeCurrentPhase() {
    guard let activeTransition = transition, let delegate = delegate else { return }
    cancelPendingPoll()

    let completion: (Result<Void, Swift.Error>) -> Void = { [weak self] result in
      guard let self else { return }
      guard let current = self.transition, current.generation == activeTransition.generation else {
        return
      }

      switch result {
      case .failure:
        self.completeActiveTransition(with: result)
      case .success:
        self.schedulePoll(generation: activeTransition.generation)
      }
    }

    switch activeTransition.phase {
    case .idle:
      completeActiveTransition(with: .success(()))
    case .waitingForInputEnable:
      delegate.issueSetInputEnabled(true, completion: completion)
    case .waitingForPublishEnable:
      delegate.issueSetPublishingEnabled(true, completion: completion)
    case .waitingForPublishDisable:
      delegate.issueSetPublishingEnabled(false, completion: completion)
    case .waitingForInputDisable:
      delegate.issueSetInputEnabled(false, completion: completion)
    }
  }

  private func schedulePoll(generation: Int64) {
    guard let delegate else { return }
    cancelPendingPoll()
    pendingPoll = delegate.schedule(delay: config.pollInterval) { [weak self] in
      self?.onPoll(generation: generation)
    }
  }

  private func onPoll(generation: Int64) {
    guard let activeTransition = transition, activeTransition.generation == generation else {
      return
    }

    if let snapshot = refreshSnapshot() {
      observeSnapshot(snapshot)
    }

    guard let current = transition, current.generation == generation else {
      return
    }

    let attempts = current.pollAttempts + 1
    if attempts < config.maxPollAttempts {
      transition = Transition(
        generation: current.generation,
        phase: current.phase,
        completion: current.completion,
        pollAttempts: attempts,
        didReapply: current.didReapply
      )
      schedulePoll(generation: generation)
      return
    }

    if !current.didReapply {
      transition = Transition(
        generation: current.generation,
        phase: current.phase,
        completion: current.completion,
        pollAttempts: 0,
        didReapply: true
      )
      executeCurrentPhase()
      return
    }

    completeActiveTransition(
      with: .failure(
        LocalMicStateControllerError(
          code: "MIC_RECONCILE_ERROR",
          message: "Daily microphone input/publishing did not reach requested state."
        )
      )
    )
  }

  private func refreshSnapshot() -> LocalMicSnapshot? {
    let snapshot = delegate?.readSnapshot()
    if let snapshot {
      observedSnapshot = snapshot
    }
    return snapshot
  }

  private func completeActiveTransition(with result: Result<Void, Swift.Error>) {
    let completion = transition?.completion
    transition = nil
    cancelPendingPoll()
    completion?(result)
  }

  private func cancelActiveTransition(with result: Result<Void, Swift.Error>) {
    let completion = transition?.completion
    transition = nil
    cancelPendingPoll()
    completion?(result)
  }

  private func cancelPendingPoll() {
    pendingPoll?.cancel()
    pendingPoll = nil
  }

  private func nextGeneration() -> Int64 {
    generationCounter += 1
    return generationCounter
  }

  private func nextPhase(snapshot: LocalMicSnapshot?, desiredEnabled: Bool) -> Phase? {
    guard let snapshot else {
      return desiredEnabled ? .waitingForInputEnable : .waitingForPublishDisable
    }

    if desiredEnabled {
      if !snapshot.inputEnabled {
        return .waitingForInputEnable
      }
      if !snapshot.publishingEnabled {
        return .waitingForPublishEnable
      }
      return nil
    }

    if snapshot.publishingEnabled {
      return .waitingForPublishDisable
    }
    if snapshot.inputEnabled {
      return .waitingForInputDisable
    }
    return nil
  }

  private func phaseSatisfied(_ phase: Phase, snapshot: LocalMicSnapshot) -> Bool {
    switch phase {
    case .idle:
      return true
    case .waitingForInputEnable:
      return snapshot.inputEnabled
    case .waitingForPublishEnable:
      return snapshot.publishingEnabled
    case .waitingForPublishDisable:
      return !snapshot.publishingEnabled
    case .waitingForInputDisable:
      return !snapshot.inputEnabled
    }
  }
}

File: pipecat_flutter_android/android/src/main/kotlin/com/kcniverba/LocalMicStateController.kt
package com.kcniverba

internal data class LocalMicSnapshot(
    val inputEnabled: Boolean,
    val publishingEnabled: Boolean,
) {
    val isSending: Boolean
        get() = inputEnabled && publishingEnabled

    fun matchesDesired(enabled: Boolean): Boolean {
        return inputEnabled == enabled && publishingEnabled == enabled
    }
}

internal data class LocalMicStateControllerConfig(
    val pollIntervalMs: Long = 100,
    val maxPollAttempts: Int = 10,
)

internal class LocalMicStateControllerError(
    val code: String,
    override val message: String,
) : RuntimeException(message)

internal interface LocalMicStateControllerDelegate {
    fun issueSetInputEnabled(enabled: Boolean, completion: (Result<Unit>) -> Unit)
    fun issueSetPublishingEnabled(enabled: Boolean, completion: (Result<Unit>) -> Unit)
    fun readSnapshot(): LocalMicSnapshot?
    fun schedule(delayMs: Long, action: () -> Unit): LocalMicScheduledTask

    /**
     * Called whenever the observed "is sending audio" state changes, including
     * poll-driven convergence that no transport callback would surface. Lets the
     * plugin re-emit the authoritative mic state to Flutter.
     */
    fun onSendingStateChanged(isSending: Boolean)
}

internal fun interface LocalMicScheduledTask {
    fun cancel()
}

internal class LocalMicStateController(
    private val delegate: LocalMicStateControllerDelegate,
    private val config: LocalMicStateControllerConfig = LocalMicStateControllerConfig(),
) {
    enum class Phase {
        IDLE,
        WAITING_FOR_INPUT_ENABLE,
        WAITING_FOR_PUBLISH_ENABLE,
        WAITING_FOR_PUBLISH_DISABLE,
        WAITING_FOR_INPUT_DISABLE,
    }

    private data class Transition(
        val generation: Long,
        val phase: Phase,
        val completion: ((Result<Unit>) -> Unit)?,
        val pollAttempts: Int,
        val didReapply: Boolean,
    )

    var desiredEnabled: Boolean = true
        private set

    var observedSnapshot: LocalMicSnapshot? = null
        private set

    val phase: Phase
        get() = transition?.phase ?: Phase.IDLE

    val isTransitioning: Boolean
        get() = transition != null

    val isSendingAudio: Boolean
        get() = observedSnapshot?.isSending == true

    private var generationCounter: Long = 0
    private var transition: Transition? = null
    private var pendingPoll: LocalMicScheduledTask? = null
    private var lastNotifiedIsSending: Boolean? = null

    fun reset(desiredEnabled: Boolean = true) {
        this.desiredEnabled = desiredEnabled
        observedSnapshot = null
        generationCounter += 1
        transition = null
        cancelPendingPoll()
        // Force the next real snapshot to notify, guaranteeing the plugin
        // re-emits the authoritative mic state after a (re)connect.
        lastNotifiedIsSending = null
    }

    fun updateDesiredEnabled(
        enabled: Boolean,
        completion: ((Result<Unit>) -> Unit)? = null,
    ) {
        desiredEnabled = enabled
        val snapshot = refreshSnapshot()
        if (snapshot?.matchesDesired(enabled) == true) {
            completeActiveTransition(Result.success(Unit))
            completion?.invoke(Result.success(Unit))
            return
        }

        cancelActiveTransition(Result.success(Unit))
        startTransition(snapshot = snapshot, completion = completion)
    }

    fun observeSnapshot(snapshot: LocalMicSnapshot) {
        observedSnapshot = snapshot
        notifySendingIfChanged()

        val activeTransition = transition
        if (activeTransition == null) {
            if (!snapshot.matchesDesired(desiredEnabled)) {
                startTransition(snapshot = snapshot, completion = null)
            }
            return
        }

        if (snapshot.matchesDesired(desiredEnabled)) {
            completeActiveTransition(Result.success(Unit))
            return
        }

        if (!phaseSatisfied(activeTransition.phase, snapshot)) {
            return
        }

        val nextPhase = nextPhase(snapshot, desiredEnabled)
        if (nextPhase == null || nextPhase == Phase.IDLE) {
            completeActiveTransition(Result.success(Unit))
            return
        }

        if (nextPhase == activeTransition.phase) {
            return
        }

        cancelPendingPoll()
        transition = activeTransition.copy(
            phase = nextPhase,
            pollAttempts = 0,
            didReapply = false,
        )
        executeCurrentPhase()
    }

    fun currentMicEnabled(fallback: Boolean): Boolean {
        return observedSnapshot?.isSending ?: fallback
    }

    private fun startTransition(
        snapshot: LocalMicSnapshot?,
        completion: ((Result<Unit>) -> Unit)?,
    ) {
        val initialPhase = nextPhase(snapshot, desiredEnabled)
        if (initialPhase == null || initialPhase == Phase.IDLE) {
            completion?.invoke(Result.success(Unit))
            return
        }

        transition = Transition(
            generation = nextGeneration(),
            phase = initialPhase,
            completion = completion,
            pollAttempts = 0,
            didReapply = false,
        )
        executeCurrentPhase()
    }

    private fun executeCurrentPhase() {
        val activeTransition = transition ?: return
        cancelPendingPoll()

        val completion: (Result<Unit>) -> Unit = completion@{ result ->
            val current = transition
            if (current == null || current.generation != activeTransition.generation) {
                return@completion
            }

            if (result.isFailure) {
                completeActiveTransition(result)
                return@completion
            }

            schedulePoll(activeTransition.generation)
        }

        when (activeTransition.phase) {
            Phase.WAITING_FOR_INPUT_ENABLE ->
                delegate.issueSetInputEnabled(enabled = true, completion = completion)

            Phase.WAITING_FOR_PUBLISH_ENABLE ->
                delegate.issueSetPublishingEnabled(enabled = true, completion = completion)

            Phase.WAITING_FOR_PUBLISH_DISABLE ->
                delegate.issueSetPublishingEnabled(enabled = false, completion = completion)

            Phase.WAITING_FOR_INPUT_DISABLE ->
                delegate.issueSetInputEnabled(enabled = false, completion = completion)

            Phase.IDLE -> {
                completeActiveTransition(Result.success(Unit))
            }
        }
    }

    private fun schedulePoll(generation: Long) {
        cancelPendingPoll()
        pendingPoll = delegate.schedule(config.pollIntervalMs) {
            onPoll(generation)
        }
    }

    private fun onPoll(generation: Long) {
        val activeTransition = transition ?: return
        if (activeTransition.generation != generation) {
            return
        }

        refreshSnapshot()?.let { observeSnapshot(it) }

        val current = transition ?: return
        if (current.generation != generation) {
            return
        }

        val attempts = current.pollAttempts + 1
        if (attempts < config.maxPollAttempts) {
            transition = current.copy(pollAttempts = attempts)
            schedulePoll(generation)
            return
        }

        if (!current.didReapply) {
            transition = current.copy(
                pollAttempts = 0,
                didReapply = true,
            )
            executeCurrentPhase()
            return
        }

        completeActiveTransition(
            Result.failure(
                LocalMicStateControllerError(
                    code = "MIC_RECONCILE_ERROR",
                    message = "Daily microphone input/publishing did not reach requested state.",
                )
            )
        )
    }

    private fun refreshSnapshot(): LocalMicSnapshot? {
        val snapshot = delegate.readSnapshot()
        if (snapshot != null) {
            observedSnapshot = snapshot
            notifySendingIfChanged()
        }
        return snapshot
    }

    private fun notifySendingIfChanged() {
        val current = observedSnapshot?.isSending == true
        if (lastNotifiedIsSending != current) {
            lastNotifiedIsSending = current
            delegate.onSendingStateChanged(current)
        }
    }

    private fun completeActiveTransition(result: Result<Unit>) {
        val completion = transition?.completion
        transition = null
        cancelPendingPoll()
        completion?.invoke(result)
    }

    private fun cancelActiveTransition(result: Result<Unit>) {
        val completion = transition?.completion
        transition = null
        cancelPendingPoll()
        completion?.invoke(result)
    }

    private fun cancelPendingPoll() {
        pendingPoll?.cancel()
        pendingPoll = null
    }

    private fun nextGeneration(): Long {
        generationCounter += 1
        return generationCounter
    }

    private fun nextPhase(
        snapshot: LocalMicSnapshot?,
        desiredEnabled: Boolean,
    ): Phase? {
        if (snapshot == null) {
            return if (desiredEnabled) {
                Phase.WAITING_FOR_INPUT_ENABLE
            } else {
                Phase.WAITING_FOR_PUBLISH_DISABLE
            }
        }

        if (desiredEnabled) {
            return when {
                !snapshot.inputEnabled -> Phase.WAITING_FOR_INPUT_ENABLE
                !snapshot.publishingEnabled -> Phase.WAITING_FOR_PUBLISH_ENABLE
                else -> null
            }
        }

        return when {
            snapshot.publishingEnabled -> Phase.WAITING_FOR_PUBLISH_DISABLE
            snapshot.inputEnabled -> Phase.WAITING_FOR_INPUT_DISABLE
            else -> null
        }
    }

    private fun phaseSatisfied(
        phase: Phase,
        snapshot: LocalMicSnapshot,
    ): Boolean {
        return when (phase) {
            Phase.IDLE -> true
            Phase.WAITING_FOR_INPUT_ENABLE -> snapshot.inputEnabled
            Phase.WAITING_FOR_PUBLISH_ENABLE -> snapshot.publishingEnabled
            Phase.WAITING_FOR_PUBLISH_DISABLE -> !snapshot.publishingEnabled
            Phase.WAITING_FOR_INPUT_DISABLE -> !snapshot.inputEnabled
        }
    }
}
